docs/fml-spec.md

```diff
@@ -0,0 +1,576 @@
+
+The "Feature Manifest" is a schema for all the features in an app. This is used for generating type safe `struct`s and `data class`es, and a schema for Experimenter to validate the configuration during experiment setup.
+
+The "Feature Manifest Language" is definition language for writing feature manifest.
+
+`nimbus-fml` is the name fo the tool for processing the feature manifest, to generate Swift and Kotlin and the JSON schema that Experimenter can ingest.
+
+Features are configured remotely by sending JSON objects to the Nimbus SDK on each device. It is the job of the `nimbus-fml`, and the code that it generates to unpack that JSON, validate it, coerce it into values usable by the app and recover if anything goes wrong.
+
+[feature-api]: feature-variables-and-me.md
+
+### JSON Merge Patch
+
+For each feature, the feature manifest should define a complete configuration needed by the app, including a default value for every variable.
+
+Remotely set experiments and rollouts will vary the configurations by presenting _patches_ on that configuration. If the complete configuration of the feature can be
+represented by a JSON object, then patches to that configuration would be applied in a manner consistent with [JSON Merge Patch RFC 7396][json-merge-patch].
+
+[json-merge-patch]: https://datatracker.ietf.org/doc/html/rfc7396
+
+### About the generated code
+
+Each project has its own manifest file, containing all the features configured in that project.
+
+Running `nimbus-fml` for a manifest generates a Swift or Kotlin class, which is named as a commmand line argument.
+
+In the examples, this class is called `FxNimbus`, but this is project specific. This is the entry point to getting values out of Nimbus.
+
+The generated class must be connected to the Nimbus SDK, which downloads the experiment recipes and decides which experiments this user is in.
+
+```swift
+// Initialize the Nimbus SDK
+let nimbus = createNimbusForApplication()
+// Connect the generated code to the Nimbus SDK
+FxNimbus.api = nimbus
+```
+
+Feature configuration for a feature individual features are accessed through the `FxNimbus.features` property. For example, a `newtab` feature's configuration can be accessed thus:
+
+```swift
+let newTabConfig = FxNimbus.features.newtab.value()
+```
+
+You should record when the user has been exposed to the feature or exposed to where the feature would've been.
+
+```swift
+FxNimbus.features.newtab.recordExposure()
+```
+
+### Identifier cases
+
+All the examples below use `kebab-case` for identifiers. When these identifiers are used to generate code, they are transformed to the language-specific casing. For example, a feature is specified in the FML as being called `spotlight-search`, but would be referred to in Swift as `spotlightSearch`.
+
+## Introducing the FML
+
+The [Feature Manifest Language is a dialect of YAML][fml-is-yaml]. The manifest file, say `nimbus.fml.yaml` must be valid YAML.
```

Comment:
It might be useful to see a complete yaml schema for the FML, that includes all possible fields and their types (I'm not sure how yaml schemas work, or if there is a standard for that, but I feel like it might make it easier to see the whole FML as one)

Reply:
Here's an example of a yaml-schema in Glean (which defines the schema for the metrics.yaml file): https://github.com/mozilla/glean_parser/blob/main/glean_parser/schemas/metrics.2-0-0.schema.yaml